src/prelude/timers.mo

```diff
@@ -0,0 +1,149 @@
+                exp := @nextExpiration(n.dopo)
+            }
+        };
+        exp
+    }
+};
+
+// Function called by backend to run eligible timed actions.
+// DO NOT RENAME without modifying compilation.
+func @timer_helper() : async () {
+    func Array_init<T>(len : Nat,  x : T) : [var T] {
+        (prim "Array.init" : <T>(Nat, T) -> [var T])<T>(len, x)
+    };
+    let now = (prim "time" : () -> Nat64)();
+    let exp = @nextExpiration @timers;
+    let prev = (prim "global_timer_set" : Nat64 -> Nat64) exp;
+
+    if (exp == 0) {
+        return
+    };
+
+    var gathered = 0;
+    let thunks = Array_init<?(() -> async ())>(10, null); // we want max 10
+
+    func gatherExpired(n : ?@Node) = switch n {
+        case null ();
+        case (?n) {
+            gatherExpired(n.ante);
+            if (n.expire > 0 and n.expire <= now and gathered < thunks.size()) {
+                thunks[gathered] := ?(n.job);
+                switch (n.delay) {
+                    case (?delay) {
+                        // re-add the node
+                        let expire = n.expire + delay;
+                        // N.B. insert only works on pruned nodes
+                        func reinsert(m : ?@Node) : @Node = switch m {
+                            case null ({ n with var expire; ante = null; dopo = null });
+                            case (?m) {
+                                assert m.expire != 0;
+                                if (expire < m.expire) ({ m with ante = ?reinsert(m.ante) })
+                                else ({ m with dopo = ?reinsert(m.dopo) })
+                            }
+                        };
+                        @timers := ?reinsert(@prune(@timers));
+                    };
+                    case _ ()
+                };
+                n.expire := 0;
+                gathered += 1;
+            };
+            gatherExpired(n.dopo);
+        }
+    };
+
+    gatherExpired(@timers);
+
+    let futures = Array_init<?(async ())>(thunks.size(), null);
+    for (k in thunks.keys()) {
+        futures[k] := switch (thunks[k]) { case (?thunk) ?thunk(); case _ null };
+    };
+
+    for (f in futures.vals()) {
+        switch f { case (?f) try await f catch _ { }; case _ () }
+    };
+};
+
+var @lastId = 0;
+
+func @setTimer(delaySecs : Nat64, recurring : Bool, job : () -> async ()) : (id : Nat) {
+    @lastId += 1;
+    let id = @lastId;
+    let now = (prim "time" : () -> Nat64) ();
+    let delayNanos = 1_000_000_000 * delaySecs;
+    let expire = now + delayNanos;
+    let delay = if recurring ?delayNanos else null;
+    // only works on pruned nodes
+    func insert(n : ?@Node) : @Node =
+        switch n {
+            case null ({ var expire; id; delay; job; ante = null; dopo = null });
+            case (?n) {
+                assert n.expire != 0;
+                if (expire < n.expire) ({ n with ante = ?insert(n.ante) })
+                else ({ n with dopo = ?insert(n.dopo) })
```

Comment:
```suggestion
                if (expire < n.expire) {
                   n.before := ?insert(n.before) 
                }
                else { 
                  n.after := ?insert(n.after) 
                }
```
With the imperative (renamed) version, you could just insert in place, without re-allocating (I think).